File: Data_Preprocessing_Meds.py
#Data Preprocessing Tools

#Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import plotly.express as px
from scipy.stats import norm



#Importing the dataset
dataset = pd.read_csv('/Users/zitachau/Downloads/HAN507/data_meds_ph.csv')


#Creating a sub-sample of the dataset since it is large/random selection of 20 rows
temp = dataset.sample(20)

#Previewing the set
temp 

#Since there is an erraneous column 1 - lets remove 
dataset = dataset.drop(columns=['Unnamed: 0'])

#Confirming that it's gone
list(dataset)

#Making sure that updates are working on the temp dataset
temp = dataset.sample(20)




#Looking what additional cleaning should happen, first up is date columns
#Transforming the date columns to actual dates
#Seeing what the column names are
list(temp)

#Appears they are 'created' and 'date_expired'
#Lets transform those to actual date/times
dataset['created'] = pd.to_datetime(dataset['created'])

#Confirming
#Updating our temp dataset, and then order that column from earliest to latest dates
temp = dataset.sample(20)
temp = temp.sort_values(by=['created'])

#Looks good
#Doing the same to the other date column 
dataset['date_expired'] = pd.to_datetime(dataset['date_expired'])

#Confirming and double checking here as well: 
temp = dataset.sample(20)
temp = temp.sort_values(by=['date_expired'])




#Seeing what other data cleaning might be warranted: 
temp 

#It appears 'name' represents drug name, lets rename the column so it is more clear
dataset = dataset.rename(columns = {'name': 'drug_name'}, inplace = False)

#Confirming change: 
temp = dataset.sample(20)
temp
#looks better!


#Doing the same for 'roa' which stands for route_of_administration 
dataset = dataset.rename(columns = {'roa': 'route_of_admin'}, inplace = False)






#For the 'frequency' column, it looks like there is the medical abbreviation, and then
#there is the actual human-readable form - lets make this one column to two columns - 
#one with the abbreviation, and a seperate one with the full description

#Making sure that the column is a string 
dataset["frequency"] = dataset["frequency"].astype(str)

#Seeing that the abbreviation and seperated by some white space, so in the code, we're
#going to give the names of the two new columns that we want to create, which is 
# "frequency_abbreviation" and "frequency_readible", then we SPLIT the columns by the white space, 
#which is designated by " " 
dataset[["frequency_abbreviation", "frequency_readible"]]= dataset["frequency"].str.split(" ", n = 1, expand = True) 


#Confirming that it worked 
temp = dataset.sample(20)
temp

#Great! But, looks like we should remove our old 'frequency' column since that is no 
#longer needed, and lets remove the ( ) from our last new variable that we created, 'frequency_readible'

dataset = dataset.drop(columns=['frequency'])
dataset['frequency_readible']= dataset['frequency_readible'].str.replace('(', '')
dataset['frequency_readible']= dataset['frequency_readible'].str.replace(')', '')

#Confirming
temp = dataset.sample(20)
temp
#looks better!


#Cleaning up the drug details column - this right now is in the structure of a DICTIONARY, but it is a string 
#First, lets convert the string to a dictionary, and then get the nested information out 
drug_details = df.drug_details.apply(eval)


#Confirming that it worked, but getting the first medispan ID value
drug_details[0]['medispan_id']

#looks good! 




#Geting the columns we want - medispan ID, NDC number, and product name
#and create this as a new dataframe
drug_details_pd = pd.DataFrame(drug_details.tolist(), columns=['medispan_id', 'ndcndf_id', 'product_name'])

#because we have the same amount of rows in our new drug_details_pd as our original dataset, 
#we're going to merge the two dataframes together into a new DF called clean_df
clean_df = df.merge(drug_details_pd, how='left', left_index=True, right_index=True)






#Understanding our missing data
df_missing = pd.DataFrame(clean_df.isnull().sum())
df_missing = df_missing.reset_index() #This resets the index, so our index now becomes a column variable that we can filter by 

#Seeing how many unique patients we have in this DF 
clean_df.UUID.nunique()

#Finding how the average number of TOTAL medications for each patient 
counts = clean_df.groupby(['UUID']).size().reset_index(name='total_med_counts') #this is grouping each row by UID - 
counts.total_med_counts.mean()

#Finding out the frequency counts for 'added_by' --> each number represents a unique physician 
added_by_counts = pd.DataFrame(clean_df.addedby.value_counts())
added_by_counts = added_by_counts.reset_index()

#Finding out what the most commonly prescribed medication is by 'product_name'
common_meds = pd.DataFrame(clean_df['product_name'].value_counts())
common_meds = common_meds.reset_index()

#Finding out which medications have the highest re-fill rate 
refilled = pd.DataFrame(clean_df.groupby(['refills_rxed','product_name']).size())
refilled = refilled.reset_index()

#Finding out which medications are the most prescribed by doses_rxed column
doses = pd.DataFrame(clean_df.groupby(['doses_rxed','product_name']).size())
doses = doses.reset_index()

#Finding out what the most common route of administration is 
roa = pd.DataFrame(clean_df['route_of_admin'].value_counts())
roa = roa.reset_index()
#Looks like there might be some repeat values in ROA, we'll take care of that later 




#Looking at some visualizations and means to understand if our continuous data is NORMAL 
#The only varialbes that are continuous that we are interested in here are doses_rxed and refills_rxed

#Getting the mean, median, and mode + quartiles for doses_rxed
#Making sure the column is not a string, but an integer 
clean_df['doses_rxed'] = pd.to_numeric(clean_df['doses_rxed'], errors='coerce')
doses_rxed_table = pd.DataFrame(clean_df.doses_rxed.describe())

#Geting the mean, median, and mode + quartiles for refills_rxed
#Making sure the column is not a string, but a integer 
clean_df['refills_rxed'] = pd.to_numeric(clean_df['refills_rxed'], errors='coerce')
refills_rxed_table = pd.DataFrame(clean_df.refills_rxed.describe())


#Making a box plot for doses_rxed 
ax = sns.boxplot(x=clean_df["doses_rxed"])

#Making a box plot for refills_rxed
ax = sns.boxplot(x=clean_df["refills_rxed"])

#Creating a histogram of doses_rxed 
sns.distplot(clean_df['doses_rxed'], fit=norm)

#Creating a histogram of doses_rxed 
sns.distplot(clean_df['refills_rxed'], fit=norm)

File: Data_Preprocessing_Meds.R
#Importing packages
library(dplyr)
library(stringr)
library(reshape2)
library(tidyverse)
library(pastecs)

#Importing the dataset
dataset = read_csv('/Users/zitachau/Downloads/HAN507/data_meds_ph.csv')

#Creating a sub-sample of the dataset since it is large/random selection of 20 rows
temp = sample_n(dataset, 20)

#Previewing the set 
temp 

#Looks like there is a erraneous column 1 - lets remove 
dataset = select(dataset, -c(1))

#Confirming that it is gone: 
colnames(dataset)

#Seeing if the updates are working on the temp dataset
temp = sample_n(dataset, 20)


#Confirming the column types in our dataset
#the below function uses sapply, which means we want to apply a function to all columns/rows in 
#our dataframe, and the argument we are using is 'class', which means we want to understand the 
#class of each of the columns in our dataframe called 'dataset'
sapply(dataset, class)

#It looks like our two dates, created and date_expired have already been processed
#as being dates, which is = POSIXct / POSIXt 
#Unlike pandas/python script, we do not need to do anything to our dates

#It appears 'name' represents drug name, lets rename the column so it is more clear
#In the line below, we say that we want to rename our 'name' column into 'drug_name'
#Unlike python, we 1) can use <- which is the equivalent of our = command, 
#and 2) that when we are renaming our column, we are not using quotations 
dataset <- dataset %>% rename(drug_name = name) 

#Confirming the change has happened: 
temp = sample_n(dataset, 20)
temp
#Looks better! 

#Doing the same for 'roa' which stands for route_of_administration 
#Below, you can see that R can still take = command, which is the same thing as <- 
dataset = dataset %>% rename(route_of_admin = roa) 

#For the 'frequency' column, it looks like there is the medical abbreviation, 
#and then there is the actual human-readable form - lets make this one column 
#two columns - one with the abbreviation, and a separate one with the full description 

#Making sure that the column is a string 
newColNames <- c("frequency_abbreviation", "frequency_readible")

#We see that the abbreviation are separate by white space, 
#Giving the names of the two new columns that we want to create, which is 
# "frequency_abbreviation" and "frequency_readible", then we SPLIT the columns by the white space, 
#which is designated by " " 
newCols <- colsplit(dataset$frequency, " ", newColNames)
dataset <- cbind(dataset, newCols)

#Confirming
colnames(dataset)
#looks good! We can see our two new columns: frequency_abbreviation and frequency_readible

#Removing our old 'frequency' column since that is no needed
dataset = subset(dataset, select = -c(frequency) )

#Confirming
temp = sample_n(dataset, 20)
temp
#looks better!



#Geting the drug details OUT of the drug_details column 
#Removing the white space using str_trim and then we will search for 
#starting and stopping words, extract those to separate lists, then merge them back together
#into our dataframe

dataset$drug_details = str_trim(dataset$drug_details, side = c("both", "left", "right"))

medispan_id <- dataset$drug_details %>% str_extract("(?<='medispan_id':).*(?='name')")
ndcndf_id <- dataset$drug_details %>% str_extract("(?<='ndcndf_id':).*(?='product_name')")
product_name <- dataset$drug_details %>% str_extract("(?<='product_name':).*(?='va_id')")

drug_details_detail <- data.frame(medispan_id=medispan_id, ndcndf_id=ndcndf_id, product_name=product_name )

clean_df <- cbind(drug_details_detail, dataset)


#Looking at a subsample to make sure it worked 
temp = sample_n(clean_df, 20)
temp
#looks better!

#Understanding our missing data
missing <- data.frame(sapply(clean_df, function(x) sum(is.na(x))))

#Seeing how many unique patients we have in this DF 
sum(!duplicated(clean_df$UUID))

#Finding the average number of TOTAL medications for each patient 
counts = data.frame(clean_df %>% group_by(UUID) %>% summarize(count=n()))
mean(counts$count)

#Finding out the frequency counts for 'added_by' --> each number represents a unique physician 
added_by_counts = data.frame(table(clean_df$addedby))

#Finding out what the most commonly prescribed medication is by 'product_name'
common_meds = data.frame(table(clean_df$product_name))

#Finding out which medications have the highest re-fill rate 
refilled = data.frame(clean_df %>% 
  group_by(refills_rxed, product_name) %>%
  summarise(number = n()))

#Finding out which medications are the most prescribed by doses_rxed column
doses = data.frame(clean_df %>% 
                        group_by(doses_rxed, product_name) %>%
                        summarise(number = n()))


#Finding out what the most common route of administration is 
roa = data.frame(table(clean_df$route_of_admin))
#Looks like there might be some repeat values in ROA, we'll take care of that later 

#Looking at visualizations and means to understand if our continuous data is NORMAL 
#The only varialbes that are continuous that we are interested in here are doses_rxed and refills_rxed

#Getting the mean, median, and mode + quartiles for doses_rxed
#Making sure the column is not a string, but a integer 
clean_df$doses_rxed = as.numeric(as.character(clean_df$doses_rxed))
doses_rxed_table = summary(clean_df$doses_rxed)


#Getting the mean, median, and mode + quartiles for refills_rxed
#Making sure the column is not a string, but an integer 
clean_df$refills_rxed = as.numeric(as.character(clean_df$refills_rxed))
refills_rxed_table = summary(clean_df$refills_rxed)


#Making box plots 
boxplot(clean_df$refills_rxed)
boxplot(clean_df$doses_rxed)

#Looking at it with a histogram
hist(clean_df$refills_rxed)
hist(clean_df$doses_rxed)
